From bd1ec441cdaf8ed76abc59daf3b43d947f9a0796 Mon Sep 17 00:00:00 2001
From: Mark Zhuang <mark.zhuang@spacemit.com>
Date: Tue, 18 Mar 2025 10:13:56 +0800
Subject: [PATCH 3/8] gcc: define __SPACEMIT_COMPILER

---
 src/gcc/cppbuiltin.cc | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/gcc/cppbuiltin.cc b/src/gcc/cppbuiltin.cc
index d0cd79cb5..f329caa89 100644
--- a/src/gcc/cppbuiltin.cc
+++ b/src/gcc/cppbuiltin.cc
@@ -73,6 +73,11 @@ define__GNUC__ (cpp_reader *pfile)
   cpp_define_formatted (pfile, "__ATOMIC_CONSUME=%d", MEMMODEL_CONSUME);
 }
 
+static void
+define__SPACEMIT__ (cpp_reader *pfile)
+{
+  cpp_define (pfile, "__SPACEMIT_COMPILER");
+}
 
 /* Define various built-in CPP macros that depend on language-independent
    compilation flags.  */
@@ -201,6 +206,7 @@ void
 define_language_independent_builtin_macros (cpp_reader *pfile)
 {
   define__GNUC__ (pfile);
+  define__SPACEMIT__ (pfile);
   define_builtin_macros_for_compilation_flags (pfile);
   define_builtin_macros_for_lp64 (pfile);
   define_builtin_macros_for_type_sizes (pfile);
-- 
2.34.1

