From: bianbu-ci <bianbu-ci@spacemit.com>
Date: Tue, 7 Apr 2026 10:45:57 +0800
Subject: Sync change from bianbu-26.04/2.43-2ubuntu1bb2

---
 string/memset.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/string/memset.c b/string/memset.c
index d345e0d..0e20caa 100644
--- a/string/memset.c
+++ b/string/memset.c
@@ -22,6 +22,20 @@
 # define MEMSET memset
 #endif
 
+#ifdef __riscv_vector
+
+void *
+inhibit_loop_to_libcall
+MEMSET (void *s, int c, size_t n)
+{
+  unsigned char *r = (unsigned char *)s, *end = (unsigned char *)s + n;
+  while (r < end)
+    *r++ = c;
+  return s;
+}
+
+#else
+
 void *
 inhibit_loop_to_libcall
 MEMSET (void *dstpp, int c, size_t len)
@@ -87,4 +101,5 @@ MEMSET (void *dstpp, int c, size_t len)
 
   return dstpp;
 }
+#endif
 libc_hidden_builtin_def (MEMSET)
